opm/parser/eclipse/EclipseState/Schedule/Events.hpp

```diff
@@ -0,0 +1,65 @@
+/*
+  Copyright 2015 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifndef SCHEDULE_EVENTS_HPP
+#define SCHEDULE_EVENTS_HPP
+
+#include <cstdint>
+#include <vector>
+
+namespace Opm
+{
+    namespace ScheduleEvents {
+        // These values are used as bitmask - 2^n structure is essential.
+        enum Events {
+            NEW_WELL = 1,
+            PRODUCTION_UPDATE = 4,
+            INJECTION_UPDATE = 8,
+            POLYMER_UPDATES = 16,
+            WELL_STATUS_CHANGE = 32,
+            COMPLETION_CHANGE = 64
+        };
+    }
+
+    /*
+      This class implements a simple system for recording when various
+      events happen in the Schedule file. The purpose of the class is
+      that downstream code can query this system whether a certain a
+      event has taken place, and then perform potentially expensive
+      calculations conditionally:
+
+        auto events = schedule->getEvents();
+        if (events(SchedulEvents::NEW_WELL , reportStep))
+           // Perform expensive calculation which must be performed
+           // when a new well is introduced.
+           ...
+
+      Querying events is done with operator().
+    */
+
+    class Events {
+    public:
+        Events();
+        void addEvent(ScheduleEvents::Events event, size_t reportStep);
+        bool operator()(ScheduleEvents::Events event, size_t reportStep) const;
```

Comment:
While somewhat cool, I think that operator() is more likely to confuse users than to inform them. I think a more (boring) traditional hasEvent() or stepHasEvent() method name would be better.